### PR TITLE
refactor(ir): remove `ops.DeferredSortKey`

### PR DESCRIPTION
**Pull request by cpcloud**
This commit removes `ops.DeferredSortKey` in favor using `Deferred`
objects. A sort direction is now rendered in every case, to be explicit
about sorting in the case that a backend has a configurable default sort
order.

BREAKING CHANGE: `ops.DeferredSortKey` is removed. Use `ops.SortKey` directly instead.

### CODE REVIEW

File: ibis/expr/deferred.py [hunk 1]
```diff
@@ -20,6 +20,11 @@ def __init__(self, resolve: Callable = toolz.identity) -> None:
         ), f"resolve argument is not callable, got {type(resolve)}"
         self.resolve = resolve
 
+    def __hash__(self) -> int:
```
Comment (kszucs) [inline ibis/expr/deferred.py:23]: I guess this is the default implementation by python, need to verify though.

Comment (cpcloud) [reply ibis/expr/deferred.py:23]: It's only the default if you leave `__eq__` and `__hash__` **both** not implemented:


```
In [1]: class Foo:
   ...:     def __eq__(self, other):
   ...:         return True
   ...:

In [2]: hash(Foo())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [2], line 1
----> 1 hash(Foo())

TypeError: unhashable type: 'Foo'
```

File: ibis/expr/api.py [hunk 2]
```diff
@@ -542,13 +544,15 @@ def asc(expr: ir.Column | str) -> ir.SortExpr | ops.DeferredSortKey:
...
     Returns
     -------
-    ir.SortExpr | ops.DeferredSortKey
-        A sort expression or deferred sort key
+    ir.ValueExpr
+        An expression
     """
-    if isinstance(expr, Expr):
-        return ops.SortKey(expr, ascending=True).to_expr()
+    if isinstance(expr, str):
```
Comment (kszucs) [inline ibis/expr/api.py:550]: Nice!

File: ibis/expr/deferred.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/deferred.py:23]: We override `__eq__` to have the expression behavior, so we have to override `__hash__` to get any kind of hashing at al.

**Review by kszucs — APPROVED** [review]
Much cleaner, thanks!

**Comment by kszucs** [comment]
@cpcloud could you please highlight the breaking change in the commit message and and `ir` as scope?

**Pull request merged**